packages/backend/ubuntu-pro.go

```diff
@@ -0,0 +1,163 @@
+package main
+import (
+    "os/exec"
+    "context"
+    "errors"
+    pb "github.com/canonical/desktop-security-center/packages/proto"
+    "google.golang.org/protobuf/types/known/emptypb"
+    "github.com/tidwall/gjson"
+    "github.com/godbus/dbus/v5"
+)
+
+var conn *dbus.Conn
+var reqId string
+
+func isServiceEnabled(basename string) (bool, error) {
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        dbus.ObjectPath("/com/canonical/UbuntuAdvantage/Services/" + basename),
+    )
+    status, err := obj.GetProperty("com.canonical.UbuntuAdvantage.Service.Status")
+    if err != nil {
+        return false, err
+    }
+    return status.Value().(string) == "enabled", nil
+}
+
+func (s *ProServer) IsMachineProAttached(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    obj := conn.Object(
+        "com.canonical.UbuntuAdvantage",
+        "/com/canonical/UbuntuAdvantage/Manager",
+    )
+    isAttached, err := obj.GetProperty("com.canonical.UbuntuAdvantage.Manager.Attached")
+    if err != nil {
+        return nil, err
+    }
+
+    return &pb.Boolean{ Ret: isAttached.Value().(bool) }, nil
+}
+
+func (s *ProServer) IsEsmInfraEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("esm_2dinfra")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+func (s *ProServer) IsEsmAppsEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("esm_2dapps")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+func (s *ProServer) IsKernelLivePatchEnabled(ctx context.Context, _ *emptypb.Empty) (*pb.Boolean, error) {
+    enabled, err := isServiceEnabled("livepatch")
+    return &pb.Boolean{ Ret: enabled }, err
+}
+
+var dbusServices map[dbus.ObjectPath]map[string]map[string]dbus.Variant
+func connectToSystemBus() error {
+    var err error
+    conn, err = dbus.ConnectSystemBus()
+    return err
+}
+
+func enableService(basename string, able string) error {
+    obj := conn.Object(
```

Comment:
I would move this to being held in the service struct and initialized in the service new function. 
If you need to reuse dbus objects (which you do here and in `attach`), it makes it easier, but the other advantage is you can ping them in the new function and fail early if your service can't reach the object for whatever reason, instead of only failing when the function is called (conn.Object does not check the object exists, only holds reference it where to find it)
[Initializing and pinging dbus object in the new function](https://github.com/matthew-hagemann/ubuntu-desktop-provision/blob/main/provd/internal/services/user/user.go#L49-L55)

Reply:
Matt, I think I've addressed your remaining comments now. I'm wondering if that's not more generalizable, could you please have a look?

Reply:
@nteodosio Hi Nathan. Sure, I'll try take a look later today :slightly_smiling_face:

Reply:
Thanks. About my concert, I think maybe it is fine, I just need to grow Manager to accommodate more services (and rename functions accordingly) as in, if I wanted to add a permissionsServer:

```
type Manager struct {
    proServer        *ProServer
    permissionServer *PermissionServer
    conn             *dbus.Conn
}

func NewServerManager (ctx context.Context) (*Manager, error) {
    conn, err := dbus.ConnectSystemBus()
    if err != nil {
        return nil, status.Errorf(codes.Internal, "Failed to connect to system bus: %s", err)
    }

    proServer, err := NewProServer(conn)
    if err != nil {
        return nil, status.Errorf(codes.Internal, "Failed to create user service: %s", err)
    }

    permissionServer, err := NewPermissionServer()
    if err != nil {
        return nil, status.Errorf(codes.Internal, "Failed to create permission server: %s", err)
    }

    return &Manager{
        proServer:        proServer,
        permissionServer: permissionServer,
        conn:             conn,
    }, nil
}
```